### PR TITLE
feat(llm): support basic rag api(v1)

### PR DESCRIPTION
**Pull request by ChenZiHong-Gavin**
There may be 2 ways to integrate fastapi into the gradio app:
1. use python client provided by Gradio
2. encapsulate asynchronous HTTP call process

refer:
1. https://www.gradio.app/guides/sharing-your-app
2. https://www.gradio.app/guides/fastapi-app-with-the-gradio-client

### CODE REVIEW

File: hugegraph-llm/src/hugegraph_llm/config/config.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ class Config:
     """HugeGraph settings"""
     graph_ip: Optional[str] = "127.0.0.1"
     graph_port: Optional[int] = 8080
-    # graph_space: Optional[str] = "DEFAULT"
+    graph_space: Optional[str] = "DEFAULT"
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/config/config.py:70]:
```suggestion
    graph_space: Optional[str] = None
```

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 2]
```diff
@@ -302,27 +366,15 @@ def embedding_settings(embedding_type):
                     ]
             else:
                 embedding_config_input = []
+
             embedding_config_button = gr.Button("apply configuration")
+            
+            # 在这里调用独立的 apply_embedding_configuration 函数
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:372]: EN comment

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 3]
```diff
@@ -433,6 +486,78 @@ def apply_embedding_configuration(arg1, arg2, arg3):
...
         btn.click(fn=init_hg_test_data, inputs=inp, outputs=out)  # pylint: disable=no-member
+    return hugegraph_llm
+
+
+def rag_web_http_api():
+    @app.post("/rag")
+    def graph_rag_api(req: RAGRequest):
+        result = graph_rag(req.query, req.raw_llm, req.vector_only, req.graph_only, req.graph_vector)
+        return {key: value for key, value in zip(
+            ["raw_llm", "vector_only", "graph_only", "graph_vector"], result) if getattr(req, key)}
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:497]: TODO: separate to another api pkg

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 4]
```diff
@@ -433,6 +486,78 @@ def apply_embedding_configuration(arg1, arg2, arg3):
...
+    @app.post("/graph/config")
+    def graph_config_api(req: GraphConfigRequest):
+        # Accept status code
+        status_code = apply_graph_configuration(req.ip, req.port, req.name, req.user, req.pwd, req.gs)
+        
+        if status_code == -1:
+            return {"message":"Unsupported HTTP method"}
+        
+        if 200 <= status_code < 300:
+            return {"message":"Connection successful. Configured finished."}
```
Comment (vichayturen) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:508]: A space after the colon should be more aesthetically pleasing

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 5]
```diff
@@ -433,6 +486,78 @@ def apply_embedding_configuration(arg1, arg2, arg3):
...
+            return {"message":f"Connection failed with status code: {status_code}"}
+        
+    @app.post("/embedding/config")
+    def embedding_config_api(req: LLMConfigRequest):
+        settings.embedding_type = req.llm_type
+
+        if req.llm_type == "openai":
+            status_code = apply_embedding_configuration(req.api_key, req.api_base, req.language_model)
+        elif req.llm_type == "qianfan_wenxin":
+            status_code = apply_embedding_configuration(req.api_key, req.api_base, None)
```
Comment (vichayturen) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:538]: At present, this function does not return a value

**Review by vichayturen — COMMENTED** [review]
Function apply_embedding_configuration does not return a value.

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 6]
```diff
@@ -141,13 +144,141 @@
...
+    host: str = None
+    port: str = None
+
+
+def test_api_connection(url, method="GET", headers=None, body=None, auth=None):
+    # TODO: use fastapi.request / starlette instead? (Also add a try-catch here)
+    response = None
+    log.debug("Request URL: %s", url)
+    if method.upper() == "GET":
+        response = requests.get(url, headers=headers, timeout=5, auth=auth)
```
[the remark this answers is not in the document]
Comment (vichayturen) [reply hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:185]: May use some methods to indirectly pass in the URL or disable this rule in the CodeQL?

**Review by vichayturen — APPROVED** [review]

File: hugegraph-llm/src/hugegraph_llm/api/rag_api.py [hunk 7]
```diff
@@ -14,3 +14,105 @@
...
+    secret_key: str = None
+    # ollama-only properties
+    host: str = None
+    port: str = None
+
+
+def rag_http_api(app: FastAPI, graph_rag_func, apply_graph_configuration_func,
+                 apply_llm_configuration_func, apply_embedding_configuration_func):
+    @app.post("/rag")
+    def graph_rag_api(req: RAGRequest):
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/api/rag_api.py:60]: Is it the proper/right way to use the fastapi?

- method -> http_method?
- class -> http_method?
- directly http_method?

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 8]
```diff
@@ -141,13 +134,110 @@ def build_kg(file, schema, example_prompt, build_mode):  # pylint: disable=too-m
         raise gr.Error(str(e))
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:134]: fixme: return 'str',  but `context` got 'dict[str, Any]'... (unify it)

image

File: hugegraph-llm/src/hugegraph_llm/api/rag_api.py [hunk 9]
```diff
@@ -14,3 +14,99 @@
...
+    # Openai-only properties
+    max_tokens: str = None
+    # qianfan-wenxin-only properties
+    secret_key: str = None
+    # ollama-only properties
+    host: str = None
+    port: str = None
+
+
+def rag_http_api(app: FastAPI, rag_answer_func, apply_graph_conf, apply_llm_conf, apply_embedding_conf):
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/api/rag_api.py:57]: what's for -> `rag_answer_func, apply_graph_conf, apply_llm_conf, apply_embedding_conf`?

File: hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py [hunk 10]
```diff
@@ -99,12 +90,13 @@ def build_kg(file, schema, example_prompt, build_mode):  # pylint: disable=too-m
...
             text += "\n"
     elif full_path.endswith(".pdf"):
+        # TODO: support PDF file
         raise gr.Error("PDF will be supported later! Try to upload text/docx now")
     else:
         raise gr.Error("Please input txt or docx file.")
     builder = KgBuilder(LLMs().get_llm(), Embeddings().get_embedding(), get_hg_client())
 
-    if build_mode != "Rebuild vertex index":
+    if build_mode != BuildMode.REBUILD_VERTEX_INDEX:
```
Comment (vichayturen) [inline hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py:99]: Maybe this should be `build_mode != BuildMode.REBUILD_VERTEX_INDEX.value`?

Comment (ChenZiHong-Gavin) [reply hugegraph-llm/src/hugegraph_llm/demo/rag_web_demo.py]: right

**Review by imbajin — APPROVED** [review]

**Pull request merged**